### PR TITLE
Refactoring the hell out of TEC

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/76729

This PR lays the base used in TEC to implement first bare bones insertion of events and linked posts.
It really makes sense when looking at TEC PR too: https://github.com/moderntribe/the-events-calendar/pull/1401

The ratio here is to move **anything we might reuse over and over** to common, general-purpose and tested methods.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | There must be a single space after the first parenthesis | [Cost_Utils.php:444](https://github.com/moderntribe/tribe-common/blob/22e19afdfb86e252294050a21c48e24f37d3db8c/src/Tribe/Cost_Utils.php#L444)
:x: | There must be a single space before the last parenthesis | [Cost_Utils.php:444](https://github.com/moderntribe/tribe-common/blob/22e19afdfb86e252294050a21c48e24f37d3db8c/src/Tribe/Cost_Utils.php#L444)
:x: | empty() and isset() must only contain variables (PHP 5.2 compatibility) | [Uploader.php:115](https://github.com/moderntribe/tribe-common/blob/22e19afdfb86e252294050a21c48e24f37d3db8c/src/Tribe/Image/Uploader.php#L115)
:x: | empty() and isset() must only contain variables (PHP 5.2 compatibility) | [Uploader.php:117](https://github.com/moderntribe/tribe-common/blob/22e19afdfb86e252294050a21c48e24f37d3db8c/src/Tribe/Image/Uploader.php#L117)
:x: | Must use &quot;self::&quot; for local static member reference | [Timezones.php:425](https://github.com/moderntribe/tribe-common/blob/22e19afdfb86e252294050a21c48e24f37d3db8c/src/Tribe/Timezones.php#L425)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
This is "on hold" as I'm not done with this yet and I might need to refactor more stuff in common while working on TEC. Take a look but do not spend any serious time reviewing this code.

File: src/Tribe/Image/Uploader.php [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		return false;
+	}
+
+	protected function maybe_init_attachment_guids_cache() {
+		if ( self::$attachment_guids_cache === false ) {
+			/** @var \wpdb $wpdb */
+			global $wpdb;
+			$guids = $wpdb->get_results( "SELECT ID, guid FROM $wpdb->posts where post_type = 'attachment'" );
+
+			self::$attachment_guids_cache = $guids ? array_combine( wp_list_pluck( $guids, 'guid' ), wp_list_pluck( $guids, 'ID' ) ) :
```
Comment (bordoni) [inline src/Tribe/Image/Uploader.php]: having only one of the parts be broken out into a new line looks a bit odd.

File: src/Tribe/Image/Uploader.php [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		if ( self::$original_urls_cache === false ) {
+			/** @var \wpdb $wpdb */
+			global $wpdb;
+			$original_urls = $wpdb->get_results( "SELECT p.ID, pm.meta_value FROM $wpdb->posts p
+					JOIN $wpdb->postmeta pm
+					ON p.ID = pm.post_id
+					WHERE p.post_type = 'attachment' AND pm.meta_key = '_tribe_importer_original_url'" );
+
+			self::$original_urls_cache = $original_urls ?
+				array_combine( wp_list_pluck( $original_urls, 'meta_value' ), wp_list_pluck( $original_urls, 'ID' ) ) : array();
```
Comment (bordoni) [inline src/Tribe/Image/Uploader.php]: having only one of the parts be broken out into a new line looks a bit odd.

File: src/Tribe/Utils/Array.php [hunk 3]
```diff
@@ -112,4 +112,33 @@ public static function strpos( $haystack, $needles, $offset = 0 ) {
...
+			return array();
+		}
+
+		if ( ! is_array( $value ) ) {
+			$value = preg_split( '/\\s*' . preg_quote( $sep ) . '\\s*/', $value );
+		}
+
+		$filtered = array();
+		foreach ( $value as $v ) {
+			if ( $v === '' ) {
```
Comment (bordoni) [inline src/Tribe/Utils/Array.php]: Yoda it UP.

**Comment by lucatume** [comment]
And this is in and fixed.

File: src/Tribe/Cost_Utils.php [hunk 4]
```diff
@@ -279,11 +288,13 @@ protected function get_cost_by_func( $costs = null, $function = 'max' ) {
...
-	 * @param $cost string Cost to parse.
+	 * @param string|array $costs        A cost string or an array of cost strings.
+	 * @param null         $max_decimals The maximum number of decimal values that should be returned in the range.
+	 * @param bool         $sort         Whether the returned values should be sorted.
 	 *
-	 * @return array
+	 * @return array An associative array of parsed costs in [ <string cost> => <cost number> ] format.
 	 */
-	public function parse_cost_range( $costs, $max_decimals = null ) {
+	public function parse_cost_range( $costs, $max_decimals = null, $sort = false ) {
```
Comment (zbtirrell) [inline src/Tribe/Cost_Utils.php]: It seems like sorting them is a good default behavior, why the reversal?

Comment (lucatume) [inline src/Tribe/Cost_Utils.php]: It really is a mistake, the param was previously named `do_not_sort` and the default stuck, good catch.

**Pull request merged**